synapse/federation/federation_client.py

```diff
@@ -358,11 +358,33 @@ async def get_pdu(
             The requested PDU, or None if we were unable to find it.
         """
 
+        logger.debug(
+            "get_pdu: event_id=%s from destinations=%s", event_id, destinations
+        )
+
         # TODO: Rate limit the number of times we try and get the same event.
 
-        ev = self._get_pdu_cache.get(event_id)
-        if ev:
-            return ev
+        event_copy = None
+        event_from_cache = self._get_pdu_cache.get(event_id)
+        if event_from_cache:
+            logger.debug("get_pdu: found event_from_cache=%s", event_from_cache)
+            assert not event_from_cache.internal_metadata.outlier, (
+                "Event from cache unexpectedly an `outlier` when it should be pristine and untouched without metadata set. "
+                "We are probably not be returning a copy of the event because downstream callers are modifying the event reference we have in the cache."
+            )
+
+            # Make sure to return a copy because downstream callers will use
+            # this event reference directly and change our original, pristine,
+            # untouched PDU. For example when people mark the event as an
+            # `outlier` (`event.internal_metadata.outlier = true`), we don't
+            # want that to propagate back into the cache.
+            event_copy = make_event_from_dict(
+                event_from_cache.get_pdu_json(),
+                event_from_cache.room_version,
+                internal_metadata_dict=None,
```

Comment:
the default for `internal_metadata_dict` is None:

```suggestion
```

Reply:
Fair enough, removed ⏩. I was trying to be more obvious why we are making a copy and what was potentially changing for the `EventBase`. But it's disconnected and not exactly relevant now that the cache is always pristine.